tools/16SDetector/hmmer-3.0/easel/miniapps/esl-alistat.man
.TH "esl-alistat" 1  "@RELEASEDATE@" "@PACKAGE@ @RELEASE@" "@PACKAGE@ Manual"

.SH NAME
.TP
esl-alistat - summarize a multiple sequence alignment file

.SH SYNOPSIS

.B esl-alistat
.I [options]
.I msafile

.SH DESCRIPTION

.pp
.B esl-alistat 
summarizes the contents of the multiple sequence alignment(s) in 
.I msafile, 
such as the alignment name, format, alignment length (number of
aligned columns), number of sequences, average pairwise % identity,
and mean, smallest, and largest raw (unaligned) lengths of the
sequences.

If 
.I msafile
is - (a single dash),
multiple alignment input is read from
.I stdin.

The alignments can be of protein or DNA/RNA sequences. All alignments
in the same 
.I msafile
must be either protein or DNA/RNA. The alphabet will be autodetected
unless one of the options 
.I --amino,
.I --dna,
or 
.I --rna 
are given. These options may be useful in automated
pipelines to make 
.B esl-alistat 
more robust; alphabet autodetection is not infallible.

The 
.B --list,
.B --icinfo,
.B --rinfo,
.B --pcinfo,
.B --psinfo,
and
.B --iinfo,
options allow dumping various statistics on the alignment to optional
output files as described for each of those options below.

The 
.B --small
option allows summarizing alignments without storing them in memory
and can be useful for large alignment files with sizes that approach
or exceed the amount of available RAM.  When
.B --small
is used, 
.B esl-alistat
will print fewer statistics on the alignment, omitting data on the
smallest and largest sequences and the average identity of the
alignment.
.B --small
only works on Pfam formatted alignments (a special type of
non-interleaved Stockholm alignment in which each sequence occurs on a
single line) and 
.BI --informat " pfam"
must be given with
.B --small.
Further, when 
.B --small
is used, the alphabet must be specified with
.I --amino,
.I --dna,
or 
.I --rna.

.SH OPTIONS

.TP
.B -h 
Print brief help;  includes version number and summary of
all options, including expert options.

.TP 
.B -1
Use a tabular output format with one line of statistics per alignment
in 
.I msafile.
This is most useful when
.I msafile
contains many different alignments (such as a Pfam database in
Stockholm format).


.SH EXPERT OPTIONS

.TP
.BI --informat " <s>"
Specify that the input alignment is in 
.I <s>
format. At present, valid formats are limited to
Stockholm, Pfam, and AFA (aligned FASTA).

.TP
.B --amino
Assert that the 
.I msafile 
contains protein sequences. 

.TP 
.B --dna
Assert that the 
.I msafile 
contains DNA sequences. 

.TP 
.B --rna
Assert that the 
.I msafile 
contains RNA sequences. 

.TP 
.B --small
Operate in small memory mode for Pfam formatted alignments.
.BI --informat " pfam"
and one of
.B --amino,
.B --dna,
or
.B --rna
must be given as well.

.TP 
.BI --list " <f>"
List the names of all sequences in the alignment to file
.I <f>.
Each name is written on its own line. Lines prefixed with "#" are
comment lines with potentially useful information about the file.

.TP 
.BI --icinfo " <f>"
Dump the information content per position in tabular format to file
.I <f>.
Lines prefixed with "#" are comment lines, which explain the
meanings of each of the tab-delimited fields.

.TP 
.BI --rinfo " <f>"
Dump information on the frequency of gaps versus nongap residues per position in tabular format to file
.I <f>.
Lines prefixed with "#" are comment lines, which explain the
meanings of each of the tab-delimited fields.

.TP 
.BI --pcinfo " <f>"
Dump per column information on posterior probabilities in tabular format to file
.I <f>.
Lines prefixed with "#" are comment lines, which explain the
meanings of each of the tab-delimited fields.

.TP 
.BI --psinfo " <f>"
Dump per sequence information on posterior probabilities in tabular format to file
.I <f>.
Lines prefixed with "#" are comment lines, which explain the
meanings of each of the tab-delimited fields.

.TP 
.BI --iinfo " <f>"
Dump information on inserted residues in tabular format to file
.I <f>.
Insert columns of the alignment are those that are gaps in the
reference (#=GC RF) annotation. This option only works if the input
file is in Stockholm format with reference annotation.
Lines prefixed with "#" are comment lines, which explain the
meanings of each of the tab-delimited fields. 

.SH AUTHOR

Easel and its documentation are @EASEL_COPYRIGHT@.
@EASEL_LICENSE@.
See COPYING in the source code distribution for more details.
The Easel home page is: @EASEL_URL@


tools/16SDetector/hmmer-3.0/easel/miniapps/esl-alistat.itest.pl
#! /usr/bin/perl

# Integrated test of the esl-alistat miniapp.
#
# Usage:     ./esl-alistat.itest.pl <esl-alistat binary> <tmpfile prefix>
# Example:   ./esl-alistat.itest.pl ./esl-alistat        foo
#
# EPN, Tue Feb  2 13:19:44 2010

$eslalistat= shift;
$tmppfx      = shift;

if (! -x "$eslalistat") { die "FAIL: didn't find esl-alistat binary $eslalistat"; }

open(ALIFILE, ">$tmppfx.stk") || die "FAIL: couldn't open $tmppfx.stk for writing alifile";
print ALIFILE << "EOF";
# STOCKHOLM 1.0
seq1         aaAAAA..AAAA...Cc.cCCCCCC.C..GGGGGgggg
#=GR seq1 PP 5789**..**88...*9.9****88.7..776543210
seq2         ..AAAAa.AAAAaacCcccCCCCCCcCccGGGGG....
#=GR seq2 PP ..********************************....
seq3         ..AAAA..AAAA...C...CCCCCC.C..GGGG-....
#=GR seq3 PP ..5555..4*44...3...888888.8..8899.....
#=GC SS_cons ...............<...<<<<......>>>>>....
#=GC PP_cons ..789*..8877...8...****99.8..99998....
#=GC RF      ..AAAA..AAAA...C...CCCCCC.c..GGGGG....
//
EOF
close ALIFILE;

open(ALIFILE, ">$tmppfx.dbl.stk") || die "FAIL: couldn't open $tmppfx.stk for writing alifile";
print ALIFILE << "EOF";
# STOCKHOLM 1.0
seq1         aaAAAA..AAAA...Cc.cCCCCCC.C..GGGGGgggg
#=GR seq1 PP 5789**..**88...*9.9****88.7..776543210
seq2         ..AAAAa.AAAAaacCcccCCCCCCcCccGGGGG....
#=GR seq2 PP ..********************************....
seq3         ..AAAA..AAAA...C...CCCCCC.C..GGGG-....
#=GR seq3 PP ..5555..4*44...3...888888.8..8899.....
#=GC SS_cons ...............<...<<<<......>>>>>....
#=GC PP_cons ..789*..8877...8...****99.8..99998....
#=GC RF      ..AAAA..AAAA...C...CCCCCC.c..GGGGG....
//
# STOCKHOLM 1.0
seq1         aaA
#=GR seq1 PP 578
seq2         ..A
#=GR seq2 PP ..*
#=GC RF      ..A
//
EOF
close ALIFILE;

open(ALIFILE, ">$tmppfx.afa") || die "FAIL: couldn't open $tmppfx.afa for writing alifile";
print ALIFILE << "EOF";
>seq1
aaAAAA..AAAA...Cc.cCCCCCC.C..GGGGGgggg
>seq2
..AAAAa.AAAAaacCcccCCCCCCcCccGGGGG....
>seq3
..AAAA..AAAA...C...CCCCCC.C..GGGG-....
EOF
close ALIFILE;

$output = `$eslalistat $smallA[$pass] -h`;
if ($? != 0)                                     { die "FAIL: esl-alistat failed unexpectedly"; }
if ($output !~ /Usage: esl-alistat/)             { die "FAIL: help output not right"; }

# We do 2 runs of most tests, with and without --small
$smallA[0] = "";
$smallA[1] = "--small --informat pfam";

for($pass = 0; $pass < 2; $pass++) {
    $pass2write = $pass+1;

    $output = `$eslalistat $smallA[$pass] --rna $tmppfx.stk 2>&1`;
    if ($? != 0)                                                                                                                                { die "FAIL: esl-alistat failed unexpectedly";}
    if ($output !~ /Alignment length:    38/)        { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($output !~ /Average length:      26.7/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($pass == 0) { 
	if ($output !~ /Format:              Stockholm/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if ($output !~ /Average identity:    93\%/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }
    else { 
	if ($output !~ /Format:              Pfam/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }

    $output = `$eslalistat $smallA[$pass] --dna $tmppfx.stk 2>&1`;
    if ($? != 0)                                                                                                                                { die "FAIL: esl-alistat failed unexpectedly";}
    if ($output !~ /Alignment length:    38/)        { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($output !~ /Average length:      26.7/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($pass == 0) { 
	if ($output !~ /Format:              Stockholm/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if ($output !~ /Average identity:    93\%/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }
    else { 
	if ($output !~ /Format:              Pfam/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }

    $output = `$eslalistat $smallA[$pass] --amino $tmppfx.stk 2>&1`;
    if ($? != 0)                                                                                                                                { die "FAIL: esl-alistat failed unexpectedly";}
    if ($output !~ /Alignment length:    38/)        { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($output !~ /Average length:      26.7/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($pass == 0) { 
	if ($output !~ /Format:              Stockholm/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if ($output !~ /Average identity:    93\%/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }
    else { 
	if ($output !~ /Format:              Pfam/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }

    if($pass == 0) { 
	$output = `$eslalistat $smallA[$pass] --informat afa --rna $tmppfx.afa 2>&1`;
	if ($? != 0)                                                                                                                               { die "FAIL: esl-alistat failed unexpectedly";}
	if ($output !~ /Format:              aligned FASTA/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if ($output !~ /Alignment length:    38/)        { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if ($output !~ /Average length:      26.7/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if ($output !~ /Average identity:    93\%/)      { die "FAIL: alignments compared incorrectly on pass $pass2write"; }
    }

    $output = `$eslalistat $smallA[$pass] -1 --rna $tmppfx.stk 2>&1`;
    if ($? != 0)                                                                                                                                { die "FAIL: esl-alistat failed unexpectedly";}
    if ($pass == 0) { 
	if ($output !~ /Stockholm       3      38           80     20     31       26.7  93/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }
    else { 
	if ($output !~ /Pfam       3      38           80       26.7/)                        { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }

    # test a file with 2 alignments
    $output = `$eslalistat $smallA[$pass] --rna $tmppfx.dbl.stk 2>&1`;
    if ($output !~ /Alignment length:    38/)        { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($output !~ /Average length:      26.7/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($output !~ /Alignment number:    2/)         { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($output !~ /Average length:      2.0/)       { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if ($pass == 0) { 
	if ($output !~ /Format:              Stockholm/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if ($output !~ /Average identity:    93\%/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }
    else { 
	if ($output !~ /Format:              Pfam/)      { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }

    # test output file options
    system("$eslalistat --icinfo $tmppfx.ic $smallA[$pass] --rna $tmppfx.dbl.stk > /dev/null");
    if ($? != 0)                                 { die "FAIL: esl-alistat failed unexpectedly";}
    $output = `cat $tmppfx.ic`;
    if($output !~ /5        9  1.0\d+  2.0\d+/)  { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if($output !~ /\# Alignment idx:  2/)        { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }

    system("$eslalistat --rinfo  $tmppfx.r  $smallA[$pass] --rna $tmppfx.dbl.stk > /dev/null");
    if ($? != 0)                                         { die "FAIL: esl-alistat failed unexpectedly";}
    $output = `cat $tmppfx.r`;
    if($output !~ /21       34         2.0  0.66\d+         1.0  0.33\d+/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    if($output !~ /\# Alignment idx:  2/)                { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }

    system("$eslalistat --pcinfo  $tmppfx.pc $smallA[$pass] --rna $tmppfx.dbl.stk > /dev/null");
    if ($? != 0)                                         { die "FAIL: esl-alistat failed unexpectedly";}
    $output = `cat $tmppfx.pc`;
    if($output !~ /31      18        3        0        0        0        0        0        0        0        1        1        0        1        0  0.82\d+/) { 
	die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; 
    }
    if($output !~ /\# Alignment idx:  2/)                { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }

    if($pass == 0) { # these are incompatible with --small 
	system("$eslalistat --psinfo  $tmppfx.ps $smallA[$pass] --rna $tmppfx.dbl.stk > /dev/null");
	if ($? != 0)                                         { die "FAIL: esl-alistat failed unexpectedly";}
	$output = `cat $tmppfx.ps`;
        if($output !~ /seq3                                           20        0        0        0        1        3        4        0        0        9        2        1  0.67\d+/) { 
	    die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; 
	}
	if($output !~ /seq2                                            1        0        0        0        0        0        0        0        0        0        0        1  0.97\d+/) {
	    die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; 
	}

	system("$eslalistat --iinfo  $tmppfx.i $smallA[$pass] --rna $tmppfx.dbl.stk > /dev/null");
	if ($? != 0)                                                  { die "FAIL: esl-alistat failed unexpectedly";}
	$output = `cat $tmppfx.i`;
        if($output !~ /16           1  0.33\d+     2.0\d+         2/) { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if($output !~ /\# Alignment idx:  2/)                         { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }

	system("$eslalistat --list $tmppfx.list $smallA[$pass] --rna $tmppfx.dbl.stk > /dev/null");
	if ($? != 0)                                                  { die "FAIL: esl-alistat failed unexpectedly";}
	$output = `cat $tmppfx.list`;
        if($output !~ /seq3/)                                         { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
	if($output !~ /\# Alignment idx:  2/)                         { die "FAIL: alignment statistics calculated incorrectly on pass $pass2write"; }
    }
}


print "ok\n"; 
unlink "$tmppfx.stk";
unlink "$tmppfx.afa";
unlink "$tmppfx.dbl.stk";
unlink "$tmppfx.i";
unlink "$tmppfx.ic";
unlink "$tmppfx.list";
unlink "$tmppfx.pc";
unlink "$tmppfx.ps";
unlink "$tmppfx.r";
exit 0;
